### ISSUE TITLE
Autoregression testing

### ISSUE THREAD

**Issue by ronoc** [mozilla/compatipede#8]
We need a way to do auto-regression testing. Of course this will be difficult for every bug. 
Ideally though it would be great to see if we could come up with some sort of DSL to define 

a) if the bug is auto testable. 
b) the target on the page where the bug can be found, CSS, screenshot, JS, markup .... 
c) the identifier for the bug, how to auto recognise it.  Botched markup, a screenshot of the bug, etc

The conductor will need to determine then on each run whether certain bugs have been fixed or not. 

Below are Hallvord's notes

What I was experimenting with for the compat-entomology project was a feature called "watch" - a table listing 1. a bug id 2. a site ID, 3. a UA family 4. some other database field (for example css_problems.property) 5. a value that if found would confirm the presence of that specific bug
So the reporting generation was along the lines of "if, when loading site A with UA X on engine Z, the field foo had value bar, site is still affected by bug 1234"
And the eventual goal was being able to enable/disable such "watch" points from the webcompat.com site - with some feature similar to this mock-up: http://hallvord.com/temp/moz/webcompat-regrUI-firstexp.png
Where the system detects and suggests data points that might be relevant for that bug, and the reviewer just confirms them with a checkbox

compat-entomology now supports these modes for data output:
# Finished - API, fields etc believed to be somewhat stable now
- /data/site.tld - big JSON dump of most recentish data about given site
  Example: 
  https://compatentomology.com.paas.allizom.org/data/rakuten.co.jp
- /diff/site.tld - checks the UA types the site recently (last 3 months) was tested with and enumerates the _differences_ from the last test results for each UA.
  Examples: 
  https://compatentomology.com.paas.allizom.org/diff/rakuten.co.jp
  https://compatentomology.com.paas.allizom.org/diff/yahoo.co.jp
- /list/list-short-name - Grabs a list of sites from arewecompatibleyet.com , diff's all recent test results
  Example:
  https://compatentomology.com.paas.allizom.org/list/jp50 

(Note: the so-called 'jp50' list is now expanded to around 130 sites, and is the only list from arewecompatibleyet.com that should have comprehensive recent test coverage in the database - however, test runs for other lists can be triggered any time..)
# Upcoming (backend should be ready but no data yet, barely tested):
- /bug/bug_id - gets the most recent test results for properties that have been reviewed and verified relevant for given bug, outputs data
- /watch/site.tld (TODO: re-name to /watch/bug_id) - URL for checking and setting what DB fields are considered relevant for a given bug.

(A few other, lower priority things are coming like /screenshots/site.tld to list what screenshots are taken lately - this data already exists in /data/ so a separate URL is just to allow clients to load less data if interested in only screenshots)
# UI ideas (webcompat.com)

Basically: http://hallvord.com/temp/moz/webcompat-regrUI-firstexp.png

What happens under the hood: a piece of JS (which is right now in an extremely rough state and doesn't use any of the Views and Models and fancy stuff webcompat.com is built from) pulls a diff/site.tld chunk of JSON data from GitHub, makes a few guesses regarding what is the "baseline" result (simple: look for "SafariMobile" ;)) and what UA is relevant for this bug report. 
All data points that are different between the "reference" UA and the "relevant" UA are assumed potentially significant for this bug, and the script generates a checkbox for each of them. Somebody looking at the bug report can enable/disable monitoring of each data point.

Whenever a checkbox is activated, we add a row to a table named "watch" listing the bug, the site and the data point we want to monitor. Going forward, the system will be able to check whether the relevant stats have changed (and do a sanity check to see if the stats have aligned between the  "reference" and the "relevant" UA). The results of this check can be shown in the bug page or end up in an entirely separate report. Or (why not?) pop up in our IRC feed suggesting we check the status of some bug :)

I also want to be able to set up "resource scan" tests easily if there's quoted source code in the bug report - but this needs to work a little differently as if the data is added to the DB we do need a way to make the DB emit the data and configure the test scripts (be they Compatipede or testsites.py or slimerjstester.js..).

(In the screenshot on http://hallvord.com/temp/moz/webcompat-regrUI-firstexp.png you also see screenshots from the testing embedded on webcompat.com - this is trivial to add if we think it adds value, since the screenshots are already sitting there.. The screenshot is of https://webcompat.com/issues/1153 if you're curious.. My ultimate goal is to have data about where in the page selectors with -webkit- problems were applied, so you can just click a specific area of the screenshot to enable "monitoring" of given CSS -webkit- issues.).

Look forward to keep tinkering with this after the vacation (feel free to start doing some UI stuff though ;)).

**Comment by dainis** [comment]
Looking at webcompat/web-bugs a lot of them are about messed up on mobile browsers layout,

for that 4 screenshots could be taken
- slimerjs with firefox UA
- slimerjs with chrome UA
- phantomjs with firefox UA
- phantomjs with chrome UA

if slimerjs screenshots with firefox and chrome UA are the same but they are different from phantomjs then that is an issue with browser engine, but if screenshots on same engine but with different UA are different then that is an issue with UA detection, if screenshots are the same accross the board then issue is with page

On some pages this comparision works great

Slimerjs firefox vs chrome user agent

![s_s_2_diff](https://cloud.githubusercontent.com/assets/946253/11365895/61ea3f3e-92b2-11e5-84d2-a51d2f0b9c01.png)

Phantomjs firefox vs chrome user agent

![p_p_2_diff](https://cloud.githubusercontent.com/assets/946253/11365898/6509007e-92b2-11e5-8da1-23b574ed8339.png)

Slimerjs vs phantomjs with firefox user agent

![p_s_g_2_diff](https://cloud.githubusercontent.com/assets/946253/11365963/cd781bc2-92b2-11e5-8d0b-b78d8344ede6.png)

Slimerjs vs phantomjs with chrome user agent

![p_s_c_2_diff](https://cloud.githubusercontent.com/assets/946253/11365968/d86d4480-92b2-11e5-92d9-4583e4bb4bc9.png)

But for some pages that have different fonts for different rendering engines results are messy

Slimerjs with firefox vs chrome user agent

![s_s_diff](https://cloud.githubusercontent.com/assets/946253/11365643/66cb42d4-92b0-11e5-9874-df690eaf7650.png)

Phantomjs with firefox vs chrome user agent

![p_p_diff](https://cloud.githubusercontent.com/assets/946253/11365658/88ba60be-92b0-11e5-8226-233ff20e17de.png)

Slimerjs vs phantomjs with firefox user agent 

![s_p_f_diff](https://cloud.githubusercontent.com/assets/946253/11365672/a4c9f4ea-92b0-11e5-96d7-f4fe17cc8525.png)

Slimerjs vs phantomjs with chrome user agent

![s_p_c_diff](https://cloud.githubusercontent.com/assets/946253/11365676/a9504d48-92b0-11e5-8d23-02b443cd687f.png)

**Comment by hallvors** [comment]
I have been investigating some JS-based approaches to pinpoint broken styling that will cause layout problems across engines more exactly. I'll polish that a bit and share shortly. Good work on the screenshot comparisons :)

**Comment by hallvors** [comment]
So here's my experiment from a couple of months ago:
JS part:
https://gist.github.com/hallvors/b0be544c17c1dac413d4
Python script (talking to Jannah instance, depends on css-analysis.js being right next to it for now)
https://gist.github.com/hallvors/064d240600f68527123c

The JS part might turn into a plugin..

**Comment by ronoc** [comment]
This feature needs to be defined more clearly before we can go ahead and start dev on this. Available to catch up at any stage.

**Comment by hallvors** [comment]
This feature needs quite some planning and defining - and probably some more experiments, actually..

However, here are some thoughts regarding a roadmap:

First, we have two "axes" - cross-browser/UA/engine comparisons, and over-time comparisons. One question is "are there interesting differences between this UA/engine and that UA/engine?" and another is "are there interesting differences between what we were served a month ago and what we're getting today?"

I'd like to do stuff more or less in this order:
1. Polish the CSS analysis code and turn it into a plugin - it's somewhat sub-optimal, would be nicer if we could access the CSS object model cross-domain through the DOM..
2. Define, review and agree between us on storage definitions (database schema, screenshot paths, whatever..) for the output data.
3. Develop a script that answers this question: "what properties might have interesting differences for site X?". For example, if you run interestingness.js with arguments site:'example.com', UA:Firefox engine:gecko it can tell you the following:
   - compared to "UA:Chrome, engine:gecko", redirects, dom_structure and screenshot have interesting differences
   - compared to "UA:Firefox, engine:webkit", there are no interesting differences
   - compared to "UA:Chrome, engine:webkit", redirects, dom_structure and screenshot have interesting differences

So basically enumerating what tests appear to be able to tell you something about the site's compatibility. (Note: it's tempting to suggest we run this analysis immediately and store the results - but that requires defining what the "reference" UA/engine is. We probably want to leave that somewhat flexible..)
4. Develop a script that flags changes for a given site between tests over time. (We probably need a "date range" argument - it's OK to default to last 4 weeks I guess. Also, for most sites out there we won't get useful data by comparing screenshots over time - most sites worth tracking will change content regularly). For example, running temporaldiff.js with site:example.com might output
- UA:Chrome, engine:gecko: redirects changed between 2015-11-05 and 2015-11-19
- UA:Firefox, engine:gecko: no changes
- UA:Chrome, engine:webkit: redirects changed between 2015-11-05 and 2015-11-19
- UA:Firefox, engine:webkit: no changes

**Comment by ronoc** [comment]
1. Polish the CSS analysis code and turn it into a plugin - it's somewhat sub-optimal, would be nicer if we could access the CSS object model cross-domain through the DOM..

@seiflotfy needs to do some exploratory work around this, after which we will know how to integrate. 
https://github.com/Asynchq/boar/issues/42

2.Define, review and agree between us on storage definitions (database schema, screenshot paths, whatever..) for the output data.
Have you looked at the couchDB schema that we have been using up until now ? It seems to me that this task is best achieved by initially running each target url against the Jannah, collecting everything that is needed to do exploratory regression testing and putting it in the db. Then after that is all done, run the various automated regression tests against this.

https://github.com/mozilla/compatipede/issues/11 
1. Develop a script that answers this question: "what properties might have interesting differences for site X?". For example, if you run interestingness.js with arguments site:'example.com', UA:Firefox engine:gecko it can tell you the following:

The plan would be that the initial test run will run all various combinations of UA and engine against each job(URI). The comparision/auto-regression testing _scripts_ should then actually be a separate process that runs afterwards, which _processes_ the results to see what it can find/ what we can learn. 
This should support a plugin type architecture to enable quick RAD of regression testing approaches.

**Comment by hallvors** [comment]
:+1: - separate "data gathering" and "analysis" steps is exactly the model I was experimenting with in the compatentomology project, and I think it's a good approach.

**Comment by hallvors** [comment]
@ronoc Looking at database contents I don't see any "plugin"-results stored anywhere. Is this work in progress?

**Comment by ronoc** [comment]
@dainis should be looking at this Monday. Mentioned here #11
